### PR TITLE
Storage Provider Interface + tests

### PR DESCRIPTION
**Pull request by wscheep**

### CODE REVIEW

File: server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+public interface MetaStoreStorageProvider {
+
+  ProtoDescriptor readDefault();
+
+  Report readShadowDelta();
+
+  void writeDefault(ProtoDescriptor defaultProtoDescriptor);
+
+  void writeShadowDelta(Report shadowDeltaReport);
```
Comment (alexvanboxel) [inline server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java]: This should not be in the provider. Provider should have a lower level interface. Too much of a responsability is pushed into the provider.

File: server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import io.anemos.metastore.core.proto.ProtoDescriptor;
+import io.anemos.metastore.v1alpha1.Report;
+
+public interface MetaStoreStorageProvider {
+
+  ProtoDescriptor readDefault();
+
+  Report readShadowDelta();
+
+  void writeDefault(ProtoDescriptor defaultProtoDescriptor);
```
Comment (alexvanboxel) [inline server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java]: Should handle standard descriptors, we're not going to expose ProtoDescriptor to the outside world.

File: server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package io.anemos.metastore.server.provider;
+
+import io.anemos.metastore.core.proto.ProtoDescriptor;
+import io.anemos.metastore.v1alpha1.Report;
+
+public interface MetaStoreStorageProvider {
+
+  ProtoDescriptor readDefault();
```
Comment (alexvanboxel) [inline server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java]: Storage should not need to know about default and shadow concept... it needs to work with a name (string)

File: server/src/main/java/io/anemos/metastore/server/MetaStore.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
+package io.anemos.metastore.server;
```
[the remark this answers is not in the document]
Comment (wscheep) [reply server/src/main/java/io/anemos/metastore/server/MetaStore.java]: this is not moved, I just added the package to reflect the module, but as you mention below this is probably overkill. I'll remove the package. Do I need to move MetaStore.java as well to somewhere else?

File: server/src/main/java/io/anemos/metastore/server/provider/GCSProvider.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import io.anemos.metastore.core.proto.ProtoDescriptor;
+import io.anemos.metastore.v1alpha1.Report;
+import java.io.IOException;
+
+public class GCSProvider implements MetaStoreStorageProvider {
```
[the remark this answers is not in the document]
Comment (wscheep) [reply server/src/main/java/io/anemos/metastore/server/provider/GCSProvider.java]: done.

File: server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java (hunk 1 above)
Comment (wscheep) [reply server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java]: refactored

File: server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java (hunk 2 above)
Comment (wscheep) [reply server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java]: refactored, however is not having any notice of descriptors but ByteStrings as there's a difference between shadow.pb and default.pb.
shadow.pb -> payload is a Report proto payload
default.pb -> payload is a FileDescriptorSet payload

File: server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java (hunk 3 above)
Comment (wscheep) [reply server/src/main/java/io/anemos/metastore/server/provider/MetaStoreStorageProvider.java]: Refactored, see comment above.

File: server/src/main/java/io/anemos/metastore/server/provider/GCSProvider.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
+package io.anemos.metastore.server.provider;
```
[the remark this answers is not in the document]
Comment (wscheep) [reply server/src/main/java/io/anemos/metastore/server/provider/GCSProvider.java]: reverted package rename

**Pull request merged**